// @apollo/gateway/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDynamicConfig = exports.isStaticConfig = exports.isPrecomposedManagedConfig = exports.isManagedConfig = exports.isManuallyManagedConfig = exports.isSupergraphSdlConfig = exports.isRemoteConfig = exports.isLocalConfig = exports.isServiceDefinitionUpdate = exports.isSupergraphSdlUpdate = void 0;
function isSupergraphSdlUpdate(update) {
    return 'supergraphSdl' in update;
}
exports.isSupergraphSdlUpdate = isSupergraphSdlUpdate;
function isServiceDefinitionUpdate(update) {
    return 'isNewSchema' in update;
}
exports.isServiceDefinitionUpdate = isServiceDefinitionUpdate;
function isLocalConfig(config) {
    return 'localServiceList' in config;
}
exports.isLocalConfig = isLocalConfig;
function isRemoteConfig(config) {
    return 'serviceList' in config;
}
exports.isRemoteConfig = isRemoteConfig;
function isSupergraphSdlConfig(config) {
    return 'supergraphSdl' in config;
}
exports.isSupergraphSdlConfig = isSupergraphSdlConfig;
function isManuallyManagedConfig(config) {
    return ('experimental_updateServiceDefinitions' in config ||
        'experimental_updateSupergraphSdl' in config);
}
exports.isManuallyManagedConfig = isManuallyManagedConfig;
function isManagedConfig(config) {
    return (isPrecomposedManagedConfig(config) ||
        (!isRemoteConfig(config) &&
            !isLocalConfig(config) &&
            !isSupergraphSdlConfig(config) &&
            !isManuallyManagedConfig(config)));
}
exports.isManagedConfig = isManagedConfig;
function isPrecomposedManagedConfig(config) {
    return ('experimental_schemaConfigDeliveryEndpoint' in config &&
        config.experimental_schemaConfigDeliveryEndpoint !== null);
}
exports.isPrecomposedManagedConfig = isPrecomposedManagedConfig;
function isStaticConfig(config) {
    return isLocalConfig(config) || isSupergraphSdlConfig(config);
}
exports.isStaticConfig = isStaticConfig;
function isDynamicConfig(config) {
    return (isRemoteConfig(config) ||
        isManagedConfig(config) ||
        isManuallyManagedConfig(config));
}
exports.isDynamicConfig = isDynamicConfig;
